modules/javafx.graphics/src/main/native-glass/mac/GlassCGLOffscreen.m

```diff
@@ -0,0 +1,302 @@
+    {
+        self->_drawCounter = 0;
+        self->_texture = 0;
+        self->_ctx = CGLRetainContext(ctx);
+
+        [self setContext];
+        {
+            self->_fbo = [[GlassCGLFrameBufferObject alloc] init];
+            if (self->_fbo == nil)
+            {
+                // TODO: implement PBuffer if needed
+                // self->_fbo = [[GlassPBuffer alloc] init];
+            }
+            [(GlassCGLFrameBufferObject*)self->_fbo setIsSwPipe:(BOOL)isSwPipe];
+        }
+        [self unsetContext];
+        if (allModes == nil) {
+            allModes = [[NSArray arrayWithObjects:NSDefaultRunLoopMode,
+                                                  NSEventTrackingRunLoopMode,
+                                                  NSModalPanelRunLoopMode, nil] retain];
+        }
+    }
+    return self;
+}
+
+- (CGLContextObj)getContext
+{
+    return self->_ctx;
+}
+
+- (void)dealloc
+{
+    if (self->_texture != 0)
+    {
+        [self bindForWidth:(GLuint)[self->glassView bounds].size.width
+                 andHeight:(GLuint)[self->glassView bounds].size.height];
+        glDeleteTextures(1, &self->_texture);
+        [self unbind];
+    }
+    [self setContext];
+    [(NSObject*)self->_fbo release];
+    self->_fbo = NULL;
+    [self unsetContext];
+
+    CGLReleaseContext(self->_ctx);
+    self->_ctx = NULL;
+
+    [super dealloc];
+}
+
+- (unsigned int)width
+{
+    return [self->_fbo width];
+}
+
+- (unsigned int)height
+{
+    return [self->_fbo height];
+}
+
+- (jlong)fbo
+{
+    return (jlong)[self->_fbo fbo];
+}
+
+- (void)setContext
+{
+    self->_ctxToRestore = CGLGetCurrentContext();
+    CGLLockContext(self->_ctx);
+    CGLSetCurrentContext(self->_ctx);
+}
+
+- (void)unsetContext
+{
+    CGLSetCurrentContext(self->_ctxToRestore);
+    CGLUnlockContext(self->_ctx);
+}
+
+- (void)bindForWidth:(GLuint)width
+           andHeight:(GLuint)height
+{
+    assert(self->_drawCounter >= 0);
+    if (self->_drawCounter == 0)
+    {
+        self->_width = width;
+        self->_height = height;
+        [self setContext];
+        [self->_fbo bindForWidth:width andHeight:height];
+    }
+    self->_drawCounter++;
+}
+
+- (void)flush:(GlassOffscreen*)glassOffScreen
+{
+    assert(self->_drawCounter > 0);
+    self->_drawCounter--;
+    if (self->_drawCounter == 0)
+    {
+        [self unbind];
+        [(GlassCGLOffscreen*)glassOffScreen blitFromOffscreen:(GlassCGLOffscreen*)self];
+        if ([NSThread isMainThread]) {
+            [[(GlassCGLOffscreen*)glassOffScreen getLayer] setNeedsDisplay];
+        } else {
+            [[(GlassCGLOffscreen*)glassOffScreen getLayer] performSelectorOnMainThread:@selector(setNeedsDisplay)
+                                                                            withObject:nil
+                                                                         waitUntilDone:NO
+                                                                                 modes:allModes];
```

Comment:
minor: this indentation is hard to look at

Reply:
Changed the indentation.